File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0, right = 0;
        int[] maxFreq = new int[26];
        int freq = 0;
        int res = 0;
        while (right < s.length()) {
            int c = s.charAt(right) - 'A';
            maxFreq[c]++;
            freq = Math.max(freq, maxFreq[c]);
            right++;
            if (right - left > freq + k) {
                int d = s.charAt(left) - 'A';
                maxFreq[d]--;
                left++;
            }
            res = Math.max(res, right - left);
        }
        return res;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         left, right = 0, 0
#         res = 0
#         window = {}
#         while right < len(s):
#             c = s[right]
#             right += 1
#             if c in window:
#                 window[c] += 1
#             else:
#                 window[c] = 1
            
#             while window[c] > 1:
#                 d = s[left]
#                 left += 1
#                 window[d] -= 1
            
#             res = max(res, right - left)
        
#         return res

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        need = set()
        left = 0
        res = 0
        for right in range(len(s)):
            while s[right] in need:
                need.remove(s[left])
                left += 1
            
            need.add(s[right])
            res = max(res, right - left + 1)
        
        return res